proxylib/cassandra/cassandraparser.go

```diff
@@ -0,0 +1,643 @@
+		ParseError("Can't get L7 rules.", rule)
+	}
+	var rules []L7NetworkPolicyRule
+	for _, l7Rule := range l7Rules.GetL7Rules() {
+		var cr CassandraRule
+		for k, v := range l7Rule.Rule {
+			switch k {
+			case "query_action":
+				cr.queryActionExact = v
+			case "query_table":
+				if v != "" {
+					cr.tableRegexCompiled = regexp.MustCompile(v)
+				}
+			default:
+				ParseError(fmt.Sprintf("Unsupported key: %s", k), rule)
+			}
+		}
+		if len(cr.queryActionExact) > 0 {
+			// ensure this is a valid query action
+			res := queryActionMap[cr.queryActionExact]
+			if res == invalidAction {
+				ParseError(fmt.Sprintf("Unable to parse L7 cassandra rule with invalid query_action: '%s'", cr.queryActionExact), rule)
+			} else if res == actionNoTable && cr.tableRegexCompiled != nil {
+				ParseError(fmt.Sprintf("query_action '%s' is not compatible with a query_table match", cr.queryActionExact), rule)
+			}
+
+		}
+
+		log.Debugf("Parsed CassandraRule pair: %v", cr)
+		rules = append(rules, &cr)
+	}
+	return rules
+}
+
+type CassandraParserFactory struct{}
+
+var cassandraParserFactory *CassandraParserFactory
+
+func init() {
+	log.Info("init(): Registering cassandraParserFactory")
+	RegisterParserFactory("cassandra", cassandraParserFactory)
+	RegisterL7RuleParser("cassandra", CassandraRuleParser)
+}
+
+type CassandraParser struct {
+	connection *Connection
+	inserted   bool
+	keyspace   string // stores current keyspace name from 'use' command
+
+	// stores prepared query string while
+	// waiting for 'prepared' reply from server
+	// with a prepared id.
+	// replies associated via stream-id
+	preparedQueryPathByStreamID map[uint16]string
+
+	// allowing us to enforce policy on query
+	// at the time of the execute command.
+	preparedQueryPathByPreparedID map[string]string // stores query string based on prepared-id,
+}
+
+func (pf *CassandraParserFactory) Create(connection *Connection) Parser {
+	log.Debugf("CassandraParserFactory: Create: %v", connection)
+
+	p := CassandraParser{connection: connection}
+	p.preparedQueryPathByStreamID = make(map[uint16]string)
+	p.preparedQueryPathByPreparedID = make(map[string]string)
+	return &p
+}
+
+func (p *CassandraParser) OnData(reply, endStream bool, dataArray [][]byte, offset int) (OpType, int) {
+
+	// inefficient, but simple for now
+	data := bytes.Join(dataArray, []byte(""))[offset:]
+	log.Debugf("OnData offset  = %d", offset)
+
+	if len(data) < cassHdrLen {
+		// Partial header received, ask for more
+		needs := cassHdrLen - len(data)
+		log.Debugf("Did not receive full header, need %d more bytes", needs)
+		return MORE, needs
+	}
+
+	// full header available, read full request length
+	requestLen := binary.BigEndian.Uint32(data[5:9])
+	log.Debugf("Request length = %d", requestLen)
+	if requestLen > cassMaxLen {
+		log.Errorf("Request length of %d is greater than 256 MB", requestLen)
+		return ERROR, 0
+	}
+
+	dataMissing := (cassHdrLen + int(requestLen)) - len(data)
+	if dataMissing > 0 {
```

Comment:
Good point, that seems like a good performance win, but I think it would complicate the parsing logic a lot, and until we get more testing with this basic form, and until we have better unit + runtime test coverage, I think it makes sense to keep the parsing logic a bit more simple.